README.md
Dump vector tiles to GeoJSON from remote URLs or local system files.

## Installation

```
npm install -g @mapbox/vt2geojson
```

## Usage

Node.js

```javascript
var vt2geojson = require('@mapbox/vt2geojson');

// remote file
vt2geojson({
    uri: 'http://api.example.com/9/150/194.mvt',
    layer: 'layer_name'
}, function (err, result) {
    if (err) throw err;
    console.log(result); // => GeoJSON FeatureCollection
});

// local file
vt2geojson({
    uri: './local/file/buffer.mvt',
    layer: 'layer_name',
    z: 9,
    x: 150,
    y: 194
}, function (err, result) {
    if (err) throw err;
    console.log(result); // => GeoJSON FeatureCollection
});
```

CLI

```
Usage: vt2geojson [options] URI

Options:
  -l, --layer  include only the specified layer
  -x           tile x coordinate (normally inferred from the URI)
  -y           tile y coordinate (normally inferred from the URI)
  -z           tile z coordinate (normally inferred from the URI)
  -h, --help   Show help  [boolean]

Examples:
  vt2geojson --layer state_label https://api.mapbox.com/v4/mapbox.mapbox-streets-v6/9/150/194.vector.pbf?access_token=${MAPBOX_ACCESS_TOKEN}
```


CHANGELOG.md
## 1.1.5

- Fix the version issues

## 1.1.4

- Return error when request for vector tiles fails, instead of throwing it [#13](https://github.com/mapbox/vt2geojson/issues/13)

## 1.1.3

- Fix GeoJSON objects returned, which included `feauture.coordinates` which is invalid according to the GeoJSON spec [#5](https://github.com/mapbox/vt2geojson/issues/5)

## 1.1.2

- Add the `vt_layer` property to features when the request consists of multiple layers [#6](https://github.com/mapbox/vt2geojson/pull/6)
- Add tests for `401` and `404` responses from a tile server

## 1.1.1

- :bug: fix TypeError when layers didn't exist in requested vector tile [#10](https://github.com/mapbox/vt2geojson/pull/10)

## 1.1.0

- Add ability to execute command on local files [#1](https://github.com/mapbox/vt2geojson/issues/1)

## 1.0.0

- First


test.js
'use strict';

var test = require('tap').test;
var nock = require('nock');
var fs = require('fs');
var vt2geojson = require('./');

var tile1 = nock('http://api.mapbox.com')
    .get('/9/150/194.mvt')
    .reply(200, fs.readFileSync('fixtures/9-150-194.mvt'));

var tile2 = nock('http://a.tiles.mapbox.com/v4/mapbox.mapbox-streets-v7')
    .get('/16/46886/30383.mvt')
    .reply(200, fs.readFileSync('fixtures/16-46886-30383.mvt'));

var invalidTokenRequest = nock('http://invalid.mapbox.com')
    .get('/16/46886/30383.mvt')
    .reply(401);

var fourohfour = nock('http://404.mapbox.com')
    .get('/16/46886/0.mvt')
    .reply(404);

test('fails without uri', function (t) {
    vt2geojson({}, function (err, result) {
       t.ok(err);
       t.equal(err.message, 'No URI found. Please provide a valid URI to your vector tile.');
       t.notOk(result);
       t.end();
    });
});

test('fails without zxy', function (t) {
    vt2geojson({
        uri: './fixtures/9-150-194.mvt',
        layer: 'state_label'
    }, function (err, result) {
       t.ok(err);
       t.equal(err.message, 'Could not determine tile z, x, and y from "./fixtures/9-150-194.mvt"; specify manually with -z <z> -x <x> -y <y>', 'expected error message');
       t.notOk(result);
       t.end();
    });
});

test('fails with 401 response: invalid token', function (t) {
  vt2geojson({
      uri: 'http://invalid.mapbox.com/16/46886/30383.mvt',
      layer: 'state_label'
  }, function (err, result) {
      t.ok(err);
      t.ok(/Invalid Token/.test(err.message), 'expected error message');
      t.end();
  });
});

test('fails 404 response', function (t) {
  vt2geojson({
      uri: 'http://404.mapbox.com/16/46886/0.mvt',
      layer: 'state_label'
  }, function (err, result) {
      t.ok(err);
      t.ok(/Error retrieving data from/.test(err.message), 'expected error message');
      t.end();
  });
});

test('url', function (t) {
    vt2geojson({
        uri: 'http://api.mapbox.com/9/150/194.mvt',
        layer: 'state_label'
    }, function (err, result) {
        t.ifError(err);
        t.deepEqual(result.type, 'FeatureCollection');
        t.deepEqual(result.features[0].properties.name, 'New Jersey');
        t.deepEqual(result.features[0].geometry, {
            type: 'Point',
            coordinates: [-74.38928604125977, 40.15027547340139]
        });
        t.end();
    });
});

test('undefined layer', function (t) {
    vt2geojson({
        uri: 'http://a.tiles.mapbox.com/v4/mapbox.mapbox-streets-v7/16/46886/30383.mvt',
        layer: 'water'
    }, function (err, result) {
        t.ifError(err);
        t.deepEqual(result.type, 'FeatureCollection');
        t.deepEqual(result.features.length, 0);
        t.end();
    });
});


test('local file: relative', function (t) {
    vt2geojson({
        uri: './fixtures/9-150-194.mvt',
        layer: 'state_label',
        z: 9,
        x: 150,
        y: 194
    }, function (err, result) {
        t.ifError(err);
        t.deepEqual(result.type, 'FeatureCollection');
        t.deepEqual(result.features[0].properties.name, 'New Jersey');
        t.deepEqual(result.features[0].geometry, {
            type: 'Point',
            coordinates: [-74.38928604125977, 40.15027547340139]
        });
        t.end();
    });
});

test('local file: absolute uri with file: protocol', function (t) {
    vt2geojson({
        uri: 'file://./fixtures/9-150-194.mvt',
        layer: 'state_label',
        z: 9,
        x: 150,
        y: 194
    }, function (err, result) {
        t.ifError(err);
        t.deepEqual(result.type, 'FeatureCollection');
        t.deepEqual(result.features[0].properties.name, 'New Jersey');
        t.deepEqual(result.features[0].geometry, {
            type: 'Point',
            coordinates: [-74.38928604125977, 40.15027547340139]
        });
        t.end();
    });
});

test('local file with directory zxy directory structure', function (t) {
    vt2geojson({
        uri: './fixtures/9/150/194.mvt',
        layer: 'state_label'
    }, function (err, result) {
        t.ifError(err);
        t.deepEqual(result.type, 'FeatureCollection');
        t.deepEqual(result.features[0].properties.name, 'New Jersey');
        t.deepEqual(result.features[0].geometry, {
            type: 'Point',
            coordinates: [-74.38928604125977, 40.15027547340139]
        });
        t.end();
    });
});

test('local file gzipped', function (t) {
    vt2geojson({
        uri: './fixtures/9-150-194.mvt.gz',
        layer: 'state_label',
        z: 9,
        x: 150,
        y: 194
    }, function (err, result) {
        t.ifError(err);
        t.deepEqual(result.type, 'FeatureCollection');
        t.deepEqual(result.features[0].properties.name, 'New Jersey');
        t.deepEqual(result.features[0].geometry, {
            type: 'Point',
            coordinates: [-74.38928604125977, 40.15027547340139]
        });
        t.end();
    });
});

test('multiple layers adds property to preserve layer name', function (t) {
    vt2geojson({
        uri: './fixtures/16-46886-30383.mvt',
        z: 16,
        x: 46886,
        y: 30383,
        layer: ['landuse', 'poi_label']
    }, function (err, result) {
        t.equal(result.type, 'FeatureCollection', 'expected type');
        t.equal(result.features.length, 112, 'expected number of features');
        result.features.forEach(function(f) {
          var val = (f.properties.vt_layer === 'landuse' || f.properties.vt_layer === 'poi_label') ? true : false;
          t.ok(val, 'expected property value');
        });
        t.end();
    });
});


index.js
'use strict';

var vt = require('vector-tile');
var request = require('request');
var Protobuf = require('pbf');
var format = require('util').format;
var fs = require('fs');
var url = require('url');
var zlib = require('zlib');

module.exports = function(args, callback) {

    if (!args.uri) return callback(new Error('No URI found. Please provide a valid URI to your vector tile.'));

    // handle zxy stuffs
    if (args.x === undefined || args.y === undefined || args.z === undefined) {
        var zxy = args.uri.match(/\/(\d+)\/(\d+)\/(\d+)/);
        if (!zxy || zxy.length < 4) {
            return callback(new Error(format("Could not determine tile z, x, and y from %s; specify manually with -z <z> -x <x> -y <y>", JSON.stringify(args.uri))));
        } else {
            args.z = zxy[1];
            args.x = zxy[2];
            args.y = zxy[3];
        }
    }

    var parsed = url.parse(args.uri);
    if (parsed.protocol && (parsed.protocol === 'http:' || parsed.protocol === 'https:')) {
        request.get({
            url: args.uri,
            gzip: true,
            encoding: null
        }, function (err, response, body) {
            if (err) {
                return callback(err);
            }
            if (response.statusCode === 401) {
                return callback(new Error('Invalid Token'));
            }
            if (response.statusCode !== 200) {
                return callback(new Error(format('Error retrieving data from %s. Server responded with code: %s', JSON.stringify(args.uri), response.statusCode)));
            }
            readTile(args, body, callback);
        });
    } else {
        if (parsed.protocol && parsed.protocol === 'file:') {
            args.uri = parsed.host + parsed.pathname;
        }
        fs.lstat(args.uri, function(err, stats) {
            if (err) throw err;
            if (stats.isFile()) {
                fs.readFile(args.uri, function(err, data) {
                    if (err) throw err;
                    readTile(args, data, callback);
                });
            }
        });
    }
};

function readTile(args, buffer, callback) {

    // handle zipped buffers
    if (buffer[0] === 0x78 && buffer[1] === 0x9C) {
        buffer = zlib.inflateSync(buffer);
    } else if (buffer[0] === 0x1F && buffer[1] === 0x8B) {
        buffer = zlib.gunzipSync(buffer);
    }

    var tile = new vt.VectorTile(new Protobuf(buffer));
    var layers = args.layer || Object.keys(tile.layers);

    if (!Array.isArray(layers))
        layers = [layers]

    var collection = {type: 'FeatureCollection', features: []};

    layers.forEach(function (layerID) {
        var layer = tile.layers[layerID];
        if (layer) {
            for (var i = 0; i < layer.length; i++) {
                var feature = layer.feature(i).toGeoJSON(args.x, args.y, args.z);
                if (layers.length > 1) feature.properties.vt_layer = layerID;
                collection.features.push(feature);
            }
        }
    });

    callback(null, collection);
}
